[Security Solution] Allow bulk upgrade rules with solvable conflicts

maximpn opened the pull request:
**Partially addresses:** https://github.com/elastic/kibana/issues/210358

## Summary
 
This PR implements functionality allowing users to bulk upgrade rules with solvable conflicts.

## Details

The main focus of this PR is to allow users to bulk upgrade rules with solvable conflicts. To achieve that the following was done

- `upgrade/_perform` dry run functionality was extended to take into account rule upgrade specifiers with resolved value
- `upgrade/_perform`'s `on_conflict` param was extended with `UPGRADE_SOLVABLE` to allow bulk upgrading rules with solvable conflicts
- UI logic updated accordingly to display rule upgrade modal when users have to make a choice to upgrade only rules without conflicts or upgrade also rules with solvable conflicts
- conflict state badges were added to the rule upgrade table

It includes changes from https://github.com/elastic/kibana/pull/213027 with some modifications.

## Screenshots

<img width="1666" alt="image" src="https://github.com/user-attachments/assets/62480830-2ed5-4278-be62-d96d53d908f6" />
<img width="1723" alt="Screenshot 2025-03-06 at 12 13 04" src="https://github.com/user-attachments/assets/b786e813-268d-49a2-80cc-81fa95d14e85" />
<img width="1724" alt="Screenshot 2025-03-06 at 12 13 30" src="https://github.com/user-attachments/assets/e5e38bd9-78a3-4026-a7ea-892bd7153938" />
<img width="1723" alt="Screenshot 2025-03-06 at 12 13 51" src="https://github.com/user-attachments/assets/d58872c3-f197-49ad-b4f3-5f45fb1efac2" />
<img width="1723" alt="Screenshot 2025-03-06 at 12 14 04" src="https://github.com/user-attachments/assets/667a6ab2-2fdb-430d-9589-1c4a6e5cdc8b" />
<img width="1722" alt="Screenshot 2025-03-06 at 12 14 17" src="https://github.com/user-attachments/assets/07f4cffe-4398-4fd5-8350-a3a2978d7dcd" />


<!--ONMERGE {"backportTargets":["8.18","8.x","9.0"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

maximpn commented:
/ci

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/281247)
* Commit: f20a17aca5f3d5aa692621e13de84b2ef88049b4

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/281247#01956809-e1b5-42e1-b417-a80b86c3aff0)

cc @maximpn

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/281247","number":281247,"commit":"f20a17aca5f3d5aa692621e13de84b2ef88049b4"}],"number":281247}
buildkite-pr-comment-->

dplumlee commented on x-pack/solutions/security/plugins/security_solution/public/detection_engine/rule_management/hooks/use_prebuilt_rules_upgrade.tsx:
```diff
@@ -111,6 +141,9 @@ export function usePrebuiltRulesUpgrade({
 
+      const { noConflictRuleIds, solvableConflictRuleIds, nonSolvableConflictRuleIds } =
+        getConflictRuleIds(ruleIds);
+
```
This can probably be combined with the logic in `conflictRuleIdsSet` but this allows us to easily use the same logic in the `ALL_RULES` function for the confirmation modal

maximpn commented:
/ci

maximpn commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/281363)
* Commit: c3888333aae30813510208d514ccfadee522e2e9

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/281363#01956b34-9476-4b85-a260-50aa8241d3c5)

### History
* :broken_heart: [Build #281247](https://buildkite.com/elastic/kibana-pull-request/builds/281247) failed f20a17aca5f3d5aa692621e13de84b2ef88049b4

cc @maximpn

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/281363","number":281363,"commit":"c3888333aae30813510208d514ccfadee522e2e9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/281247","number":281247,"commit":"f20a17aca5f3d5aa692621e13de84b2ef88049b4"}],"number":281363}
buildkite-pr-comment-->

maximpn commented:
/ci

maximpn commented:
/ci

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/281384)
* Commit: 1d35bd48824bdd4cc4f8017eb144130c3092e7cc

### Failed CI Steps
* [FTR Configs #30](https://buildkite.com/elastic/kibana-pull-request/builds/281384#01956b6f-188a-403c-8686-99a5f26f1a87)
* [FTR Configs #30](https://buildkite.com/elastic/kibana-pull-request/builds/281384#01956b8c-7759-452b-97e7-d90f7e184487)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/281384#01956b8c-7759-452b-97e7-d90f7e184487) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/281384/jobs/01956b8c-7759-452b-97e7-d90f7e184487/artifacts/01956b97-58c3-4667-b0ac-d78e6cac39d2) FTR Configs #<span></span>30 / lens app - group 5 lens drag and drop tests keyboard drag and drop should move dimension to incompatible dimension
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/281384#01956b6f-188a-403c-8686-99a5f26f1a87) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/281384/jobs/01956b6f-188a-403c-8686-99a5f26f1a87/artifacts/01956b8c-4e2a-430f-86ee-4d950c5252bb) FTR Configs #<span></span>30 / lens app - group 5 lens drag and drop tests keyboard drag and drop should reorder elements with keyboard

### History
* :broken_heart: [Build #281363](https://buildkite.com/elastic/kibana-pull-request/builds/281363) failed c3888333aae30813510208d514ccfadee522e2e9
* :broken_heart: [Build #281247](https://buildkite.com/elastic/kibana-pull-request/builds/281247) failed f20a17aca5f3d5aa692621e13de84b2ef88049b4

cc @maximpn

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/281384","number":281384,"commit":"1d35bd48824bdd4cc4f8017eb144130c3092e7cc"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/281363","number":281363,"commit":"c3888333aae30813510208d514ccfadee522e2e9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/281247","number":281247,"commit":"f20a17aca5f3d5aa692621e13de84b2ef88049b4"}],"number":281384}
buildkite-pr-comment-->

dplumlee reviewed (approved):
LGTM, pulled down and tested all logic paths for upgrading different rule types from flyout and table

kibanamachine commented:
Starting backport for target branches: 8.18, 8.x, 9.0

https://github.com/elastic/kibana/actions/runs/13704789457

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/213448">](https://github.com/elastic/kibana/pull/213448)|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/213449">](https://github.com/elastic/kibana/pull/213449)|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/213450">](https://github.com/elastic/kibana/pull/213450)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

maximpn commented:
Hi @nastasha-solomon,

we added conflict badges to rule upgrade table and changed rule upgrade confirmation modal (you can see the changes in the screenshots in the PR description). Could you have a look and give a feedback?

nastasha-solomon commented:
Hey, @maximpn - did you need feedback on something in particular? Since these changes have already been merged, I want to make sure that I'm providing useful and targeted feedback.

nastasha-solomon commented:
Hey @maximpn - last Friday @dplumlee and I went over these changes, so I was able to get some additional context for the copy review. I placed my copy suggestions [here](https://docs.google.com/document/d/1qjSxs1aDHJ8AKj5xR4erPdNumLy9Jqu3AlL9oGXnfxo/edit?usp=sharing) to avoid adding more comments to this already closed PR. Feel free to post in the doc if you have questions or suggestions.

cc: @banderror @approksiu
